### ISSUE TITLE
Former 'om revert-staged-changes' has no equivalent, as 'om unstage-product' actually plans a deletion of the product

### ISSUE THREAD

**Issue by bgandon** [pivotal-cf/om#421]
Hi,

In our Concourse automation pipelines that manage our sandbox, staging, and production PCF foundations, we were using the former `om revert-staged-changes` right before exporting Ops Man settings with `om export-installation`.

Exporting Ops Man settings is part of our backup plan, and we need it especially before actually upgrading Ops Man (which is a re-creation of the VM with a new image, followed by a re-import of the previously saved installation settings).

With `om` version 3.0.0, there is no more `om revert-staged-changes`. The [release notes](https://github.com/pivotal-cf/om/releases/tag/3.0.0) suggest that `om unstage-product` does the same, so we immediately implemented a replacement for `om revert-staged-changes` based on `om unstage-product`.

We were not far from a big outage when we saw that `om unstage-product` actually plans a deletion of the targeted product! So obviously `om unstage-product` is not what we need, when trying to put Ops Man back to its last known “clean” state.

So here are our questions:
1. Could you update the `om` CLI documentation in order to provide details for the `unstage-product` command?
2. There is a “Revert” button that appears in the _Installation Dashboard_ when some changes are pending. This is what we need to achieve with the `om` CLI in our Concourse automation. Can you suggest a replacement for `om revert-staged-changes` that would do the same as this “Revert” button  ?

Thank you for your answers!

Best,
Benjamin

**Comment by kcboyle** [comment]
@bgandon 
There are no plans in the immediate present to add this support back into `om`. 
However, the team has talked to the ops manager team about potentially getting this API endpoint added. That said, if we can get the API support, there are no guarantees that this can backported to older versions of Ops Manager. 

What kind of documentation changes are you proposing? We would be happy to add that documentation in.

**Comment by kcboyle** [comment]
@bgandon do you have any documentation changes we can make as it stands at present?

We've started a conversation with the ops manager team to add revert-staged-changes to the API, which could then be supported in `om`, but there is nothing definite yet.

**Comment by bgandon** [comment]
As far as documentation could be improved, please first remove this sentence from the [v3.0.0 release notes](https://github.com/pivotal-cf/om/releases/tag/3.0.0): “_`unstage-product` functionally does the same thing_” because it is not true. If I understand it correctly, it looks like “_there is no replacement yet_” would be better there.

Then, our main concern is to update our Concourse automation in order to have it do what it is supposed to do (i.e. export OpsMan settings before any backup or OpsMan upgrade) and still use recent versions of the `om` CLI.

Our pipeline automation, as developed in a Pivotal Dojo last year, requires this `om revert-staged-changes` before running `om export-installation`. At the time, the Pivotal Solution Architect told us it was required. But is this really necessary? We don't know. Currently, as a workaround, we've pinned our PCF pipeline image to v2.0.1 of `om`, in order to still have support for `om revert-staged-changes`.

We would like to know what are the consequences if we run `om export-installation` with pending changes that would not be cancelled by `om revert-staged-changes`. Would this be bad or harmless? (We are unfortunately unclear about what `export-installation` does, and what are its prerequisites.)

If you can give some insights it would be great. This could possibly unblock us, whenever `om revert-staged-changes` turns out to be unnecessary in this use-case. Or it could lead to a documentation improvement, would `om revert-staged-changes` turn out to be a prerequisite to exporting OpsMan installation settings.

**Comment by kcboyle** [comment]
Hi @bgandon! We update the release notes to say something more like the following:
```
remove revert-staged-changes unstage-product will revert the changes if the tile has not been installed. There is currently no replacement for this command, however, it was not working for newer versions of Ops Manager, and did nothing.
```

I would definitely recommend the workflow you described. `export-installation` is the most important part of the backup/upgrade process. However, it does not require a `revert-staged-changes` before running. 

`export-installation` will only export the deployed state of the ops manager (if you try to run it before installing for the first time, it will actually fail, for instance). If you have anything pending that has not been deployed, this will *not* be backed up in the exported installation.

The only time the pending changes could potentially block your workflow is if you needed a clean slate before doing a configuration or an `apply-changes`. However, if you have exported your configuration using `staged-config` or `staged-director-config`, you can use that configuration as the source of truth for keeping your foundation's configuration up-to-date. the `pending-changes` command can help you to narrow down if there are unexpected changes in your foundation, and, if necessary, can be reverted in the UI. 

In the meantime, the team is still trying to make an effort to encourage the addition of adding the `revert` functionality to the API. Until then, `om` will be unable to do so.

**Comment by bgandon** [comment]
Thank you for these clarifications. It's fine for us, to I suggest we close this issue.

**Comment by kcboyle** [comment]
We can close this issue, or we are happy to keep it open with a "discussion" tag to keep you posted on the progress of adding it into ops man if you would like. I'll leave the decision up to you, @bgandon on whether you would like to close it.

**Comment by tonytheleg** [comment]
To add to the argument of getting revert-staged-changes back, we had a pipeline that would ensure we didn't have configuration drift but grabbing each tiles manifests and om config, then configure each product using our git-versioned properties files, and then grab those manifests and such again and tattle tale when we made changes to tiles and not our configuration files. Not being able to revert changes after done could potentially create more issues.

**Comment by jtarchie** [comment]
There is no argument of getting it back. I think it is getting lost here, the command didn't actually work. It existed, and at one time worked for older versions os Opsman, but on 2.2+ it didn't work. It just didn't have a failure message, so it always looked like it worked, but silently failing.

That being said, we are working with the OpsManager team to add the API endpoint, to bring this feature into the _future_.
We aren't ignoring this feature, we just actually want it to work. Not pretend to work. 😄

**Comment by tonytheleg** [comment]
I think I'm ok with it pretending to work as long as there is some form of random daily affirmation message with it...like
`FAILED: revert-staged-changes doesn't actually work...but you sure are looking sharp today, wink wink fingerguns` :)

All kidding aside (clearly I skimmed instead of reading the entire convo), thanks for the update and all you guys do. Om has been huge in the way we manage and automate our platform long before the Platform Automation GA and back when there was no selective applies in Ops Manager. Its an incredible tool and I'm happy to see you all really cranking out more and more features with it. You are all killing it and I thank you for making this tool!

**Comment by jtarchie** [comment]
The OpsManagers of 2.5+ have released the endpoint to support this feature.
We are now starting on adding it to `om`.

**Comment by jtarchie** [comment]
Released in [4.3](https://github.com/pivotal-cf/om/releases/tag/4.3.0). Please note the Ops Manager versions required.
